oxidation/libparsec/crates/types/src/regex.rs

```diff
@@ -12,43 +30,137 @@ fn escape_globing_pattern(string: &str) -> String {
 }
 
 impl Regex {
-    pub fn from_pattern(pattern: &str) -> Result<Self, fnmatch_regex::error::Error> {
-        let escaped_str = escape_globing_pattern(pattern);
-        Ok(Regex(fnmatch_regex::glob_to_regex(&escaped_str)?))
+    /// Returns a regex which is built from a file that contains shell like patterns
+    pub fn from_file(file_path: &Path) -> RegexResult {
+        let file_content =
+            fs::read_to_string(file_path).map_err(|err| RegexError::PatternFileIOError { err })?;
+        let mut regexes = Vec::new();
+        for pattern_line in file_content.lines() {
+            // Skip comment and empty line
+            if pattern_line.trim().starts_with('#') || pattern_line.trim() == "\n" {
+                continue;
+            }
+
+            let escaped_glob = escape_globing_pattern(pattern_line.trim());
+            regexes.push(
+                fnmatch_regex::glob_to_regex(&escaped_glob)
+                    .map_err(|err| RegexError::ShellPatternError { err })?,
+            );
+        }
+
+        Ok(Self(regexes))
     }
 
-    pub fn from_regex_str(regex_str: &str) -> Result<Self, regex::Error> {
-        Ok(Regex(regex::Regex::new(regex_str)?))
+    /// Returns a regex which is an union of all regexes from `raw_regexes` slice parameter
+    pub fn from_raw_regexes(raw_regexes: &[&str]) -> RegexResult {
+        let mut regexes = Vec::new();
+        for regex_str in raw_regexes {
+            regexes
+                .push(regex::Regex::new(regex_str).map_err(|err| RegexError::ParseError { err })?)
+        }
+
+        Ok(Self(regexes))
     }
 
-    pub fn is_match(&self, string: &str) -> bool {
-        self.0.is_match(string)
+    pub fn from_pattern(pattern: &str) -> RegexResult {
+        let escaped_str = escape_globing_pattern(pattern);
+        Ok(Regex(vec![fnmatch_regex::glob_to_regex(&escaped_str)
+            .map_err(|err| RegexError::ShellPatternError { err })?]))
     }
-}
 
-impl AsRef<str> for Regex {
-    fn as_ref(&self) -> &str {
-        self.0.as_str()
+    pub fn from_regex_str(regex_str: &str) -> RegexResult {
+        Ok(Regex(vec![
+            regex::Regex::new(regex_str).map_err(|err| RegexError::ParseError { err })?
+        ]))
+    }
+
+    pub fn is_match(&self, string: &str) -> bool {
+        self.0.iter().any(|r| r.is_match(string))
     }
 }
 
 impl PartialEq for Regex {
     fn eq(&self, other: &Self) -> bool {
-        self.as_ref() == other.as_ref()
+        if self.0.len() == other.0.len() {
+            let lhs: Vec<&str> = self.0.iter().map(|r| r.as_str()).collect();
+            let rhs: Vec<&str> = other.0.iter().map(|r| r.as_str()).collect();
+
+            lhs == rhs
```

Comment:
I don't find this comparison good:

1. We're doing allocation
2. How we should handle `[1, 2, 3]` vs `[1, 3, 2]` ? (i.e. item may not be in the same order between 2 Regex)

Reply:
I've fix the order issue with a `HashSet`. It should also do less allocations. Do we need regex comparison outside of integration test?

Reply:
I dunnon, we can disable `PartialEq` outside of test by adding `#[cfg(test)]` above the `impl PartialEq`

Reply:
Yes, but the overload is needed for pytest, `cfg(test)` is not active when we do `maturin develop` so regex comparisons will always return `False` in python tests.